src/NetTopologySuite/Operation/OverlayNG/PrecisionUtility.cs

```diff
@@ -0,0 +1,350 @@
+        /// to determine the number of decimal places,
+        /// This is likely not optimal for performance, 
+        /// but should be accurate and portable. 
+        /// </summary>
+        /// <param name="value">A numeric value</param>
+        /// <returns>The number of decimal places in the value</returns>
+        private static int NumberOfDecimals(double value)
+        {
+            /*
+             * Ensure that scientific notation is NOT used
+             * (it would skew the number of fraction digits)
+             */
+            string s = OrdinateFormat.Default.Format(value);
+            if (s.EndsWith(".0"))
+                return 0;
+            int len = s.Length;
+            int decIndex = s.IndexOf('.');
+            if (decIndex <= 0)
+                return 0;
+            return len - decIndex - 1;
+        }
+
+        /// <summary>
+        /// Applies the inherent scale calculation 
+        /// to every ordinate in a geometry.
+        /// <para/>
+        /// WARNING: this is <b>very</b> slow.
+        /// </summary>
+        /// <author>Martin Davis</author>
+        private class InherentScaleFilter : IEntireCoordinateSequenceFilter
+        {
+
+            private double _scale = 0;
+
+            public double Scale
+            {
+                get => _scale;
+            }
+
+            public void Filter(CoordinateSequence sequence)
+            {
+                for (int i = 0; i < sequence.Count; i++)
+                {
+                    UpdateScaleMax(sequence.GetX(i));
+                    UpdateScaleMax(sequence.GetY(i));
+                }
+            }
+
+            private void UpdateScaleMax(double value)
+            {
+                double scaleVal = PrecisionUtility.InherentScale(value);
+                if (scaleVal > _scale)
+                {
+                    //System.out.println("Value " + value + " has scale: " + scaleVal);
+                    _scale = scaleVal;
+                }
+            }
+
+            public bool Done
+            {
+                get => false;
+            }
+
+            public bool GeometryChanged
+            {
+                get => false;
+            }
+        }
+
+        /// <summary>
+        /// Determines a precision model to
+        /// use for robust overlay operations for one geometry.
+        /// The precision scale factor is chosen to maximize
+        /// output precision while avoiding round-off issues.
+        /// <para/>
+        /// NOTE: this is a heuristic determination, so is not guaranteed to
+        /// eliminate precision issues.
+        /// <para/>
+        /// WARNING: this is <b>very</b> slow.
+        /// </summary>
+        /// <param name="a">A geometry</param>
+        /// <returns>A suitable precision model for overlay</returns>
+        public static PrecisionModel RobustPM(Geometry a)
+        {
+            double scale = PrecisionUtility.RobustScale(a);
```

Comment:
validate `a` non-null before using it

(we could rely on `RobustScale` to validate for us if we do an earlier comment, mainly just mentioning this here for consistency)